x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/multiterms/MultiTermsAggregator.java

```diff
@@ -0,0 +1,538 @@
+            List<TermValues> termValuesList = termValuesList(ctx);
+            for (int docId = 0; docId < ctx.reader().maxDoc(); ++docId) {
+                termsIterator(docTerms(termValuesList, docId), owningBucketOrd, l -> {});
+            }
+        }
+    }
+
+    @Override
+    public InternalAggregation buildEmptyAggregation() {
+        return new InternalMultiTerms(
+            name,
+            order,
+            order,
+            bucketCountThresholds.getRequiredSize(),
+            bucketCountThresholds.getMinDocCount(),
+            bucketCountThresholds.getShardSize(),
+            showTermDocCountError,
+            0,
+            emptyList(),
+            0,
+            formats,
+            keyConverters,
+            metadata()
+        );
+    }
+
+    static TermValuesSource buildNumericTermValues(ValuesSourceConfig config) {
+        final ValuesSource.Numeric vs = (ValuesSource.Numeric) config.getValuesSource();
+        if (vs.isFloatingPoint()) {
+            return new DoubleTermValuesSource(config);
+        } else {
+            return new LongTermValuesSource(config);
+        }
+    }
+
+    /**
+     * Capture type-specific functionality of each term that comprises the multi_term key
+     */
+    interface TermValuesSource {
+        /**
+         * used in getLeafCollector to obtain a doc values for the given type
+         */
+        TermValues getValues(LeafReaderContext ctx) throws IOException;
+
+        /**
+         * Returns a key converter that knows how to convert key values into user-friendly representation and format them as a string
+         */
+        InternalMultiTerms.KeyConverter keyConverter();
+    }
+
+    interface TermValues {
+        /**
+         * Returns a list of values retrieved from doc values for the given document
+         */
+        List<Object> collectValues(int doc) throws IOException;
+    }
+
+    /**
+     * Handles non-float and date doc values
+     */
+    static class LongTermValuesSource implements TermValuesSource {
+        ValuesSource.Numeric source;
+        InternalMultiTerms.KeyConverter converter;
+
+        LongTermValuesSource(ValuesSourceConfig config) {
+            this.source = (ValuesSource.Numeric) config.getValuesSource();
+            if (config.format() == DocValueFormat.UNSIGNED_LONG_SHIFTED) {
+                converter = InternalMultiTerms.KeyConverter.UNSIGNED_LONG;
+            } else {
+                converter = InternalMultiTerms.KeyConverter.LONG;
+            }
+        }
+
+        @Override
+        public TermValues getValues(LeafReaderContext ctx) throws IOException {
+            SortedNumericDocValues values = source.longValues(ctx);
+            return doc -> {
+                if (values.advanceExact(doc)) {
+                    List<Object> objects = new ArrayList<>();
+                    int valuesCount = values.docValueCount();
+                    long previous = Long.MAX_VALUE;
```

Comment:
what happens if the last value really was Long.MAX_VALUE?  we accept that as an indexable value rigt?

Reply:
Do you mean the first value (not the last value)? The first value is handled a 3 lines below in 
``` 
if (previous != val || i == 0) {
```
The last value shouldn't be an issues unless I am missing something. I think the initial assignment here is not important (it's there to just make the compiler happy). This code was heavily inspired by the terms aggs and didn't see reason to deviate from the terms agg implementation here.